File: week1/2/A.py
# https://leetcode.com/problems/implement-stack-using-queues/
# https://zhenyu0519.github.io/2020/07/03/lc255/#methodology

import collections

class MyStack:

    def __init__(self):
        self._queue= collections.deque()

    def push(self, x: int) -> None:
        self._queue.append(x)
        for _ in range(len(self._queue) -1):
            self._queue.append(self._queue.popleft())

    def pop(self) -> int:
            return self._queue.popleft()

    def top(self) -> int:
        return self._queue[0]

    def empty(self) -> bool:
        return len(self._queue) == 0

File: week1/2/A.rs
// https://github.com/doocs/leetcode/blob/main/solution/0200-0299/0225.Implement%20Stack%20using%20Queues/README_EN.md

use std::collections::VecDeque;

struct MyStack{
    q_1: VecDeque<i32>,
    q_2: VecDeque<i32>,
    index: i32,
}

impl MyStack {
    fn new() -> Self {
        Self{
            q_1: VecDeque::new()
            q_2: VecDeque::new()
            index: 1_i32,
        }
    }

    fn move_data(&mut self){
        while !self.q_1.is_empty(){
            self.q_2.push_back(self.pop_front().unwrap());
        }
        let tmp= self.q_1.clone();
        self.q_1= self.q_2.clone();
        self.q_2= tmp;
    }
    
    fn push(&mut self, x: i32) {
        self.q_2.push_back(x);
        self.move_data();
    }

    fn pop(&mut self) -> i32 {
        self.q_1.pop_front().unwrap()
    }

    fn top(&mut self) -> i32 {
        *self.q_1.front().unwrap()
    }

    fn empty(&self) -> bool {
        self.q_1.is_empty()
    }
}